<!-- frobit launch file 

FroboMind controller frobits using 22000/3300 ohm voltage dividers:
	supply_voltage_scale_factor = 0.03747

RoboCard based frobits using 1800/700 ohm voltage dividers: 
	supply_voltage_scale_factor = 0.01746

-->
<launch>
	<!-- Are we using simulation? -->
	<arg name="sim" default="true"/>
	<param name="use_sim_time" value="true" if="$(arg sim)" />
	
	<!-- Static transforms -->
	<node pkg="tf" type="static_transform_publisher" name="imuToLink" args=" 0 0 0 0 0 3.14 base_link imu_link 20" />
	<node pkg="tf" type="static_transform_publisher" name="laserToLink" args="0.265 0 0 0 0 0 base_link laser_mount_link 20" />

	<group ns="fmLib" unless="$(arg sim)">
		<node pkg="sdu_frobit_v2" name="frobit" type="frobit_interface_node.py" output="screen" >
			<param name="deadman_sub" value="/fmCommand/deadman" />
			<param name="cmd_vel_sub" value="/fmCommand/cmd_vel" />
			<param name="enc_left_pub" value="/fmInformation/encoder_left" />
			<param name="enc_right_pub" value="/fmInformation/encoder_right" />
			<param name="wheel_status_left_pub" value="/fmInformation/wheel_status_left" />
			<param name="wheel_status_right_pub" value="/fmInformation/wheel_status_right" />
			<param name="wheel_feedback_left_pub" value="/fmInformation/wheel_feedback_left" />
			<param name="wheel_feedback_right_pub" value="/fmInformation/wheel_feedback_right" />
			<param name="wheel_pid_left_pub" value="/fmInformation/wheel_pid_left" />
			<param name="wheel_pid_right_pub" value="/fmInformation/wheel_pid_right" />
			<param name="nmea_from_frobit_sub" value="/fmSignal/nmea_from_frobit" />	 
			<param name="nmea_to_frobit_pub" value="/fmSignal/nmea_to_frobit" />	 
			<param name="publish_wheel_status_rate" value="10" />
			<param name="publish_wheel_feedback_rate" value="10" />
			<param name="publish_wheel_pid_rate" value="10" />
			<param name="castor_front" value="true"/>
			<param name="max_linear_acceleration" value="16.2"/>
			<param name="max_angular_acceleration" value="16.0"/>
			<param name="wheel_drive_feed_forward" value="85" /> 
			<param name="wheel_drive_kp" value="5" /> 
			<param name="wheel_drive_ki" value="10" /> 
			<param name="wheel_drive_kd" value="5" /> 
			<param name="wheel_drive_max_integral_output" value="300" /> 
			<param name="wheel_turn_feed_forward" value="110" /> 
			<param name="wheel_turn_kp" value="10" /> 
			<param name="wheel_turn_ki" value="20" /> 
			<param name="wheel_turn_kd" value="3" /> 
			<param name="wheel_turn_max_integral_output" value="300" /> 
			<param name="supply_voltage_scale_factor" value="0.03747" /> 
			<param name="min_supply_voltage" value="12.1" /> 
			<param name="show_voltage_interval" value="300" /> 
		</node>

		<!-- frobit serial device and protocol interface -->
		<node pkg="serial_nmea" name="frobit_nmea" type="serial_nmea_node" respawn="true" output="screen">
			<param name="nmea_from_device_pub" value="/fmSignal/nmea_from_frobit"/>
			<param name="nmea_to_device_sub" value="/fmSignal/nmea_to_frobit"/>
			<param name="serial_device" value="/dev/frobit"/>
			<param name="serial_baudrate" value="57600"/>
			<param name="use_nmea_checksum" value="true"/>
		</node>
        </group>
        <group ns="fmProcessors" unless="$(arg sim)">
		<!-- calculate odometry based on wheel encoders and IMU yaw axis orientation -->
		<node pkg="differential_odometry" name="differential_odometry" type="differential_odometry_node" output="screen" >
			<param name="enc_left_sub" value="/fmInformation/encoder_left" />
			<param name="enc_right_sub" value="/fmInformation/encoder_right" />
			<param name="imu_sub" value="/fmInformation/imu" />
			<param name="odom_reset_sub" value="/fmInformation/odom_reset" />
			<!--param name="odom_pub" value="/fmKnowledge/odom" /-->
			<param name="odom_pub" value="/fmKnowledge/pose" />
			<param name="odom_estimate_frame" value="/odom" />
			<param name="vehicle_frame" value="base_footprint" />
			<param name="encoder_output" value="absolute" />
			<param name="max_ticks_per_update" value="200" />
			<param name="yaw_angle_source" value="imu_angular_velocity" />
			<param name="imu_angular_velocity_yaw_axis" value="z" />
			<param name="diff_steer_wheel_distance" value="0.37"/>
			<param name="ticks_per_meter_left" value="710"></param>  
			<param name="ticks_per_meter_right" value="710"></param>  
		</node>
	</group>  
        <group if="$(arg sim)">
		<node pkg="stage_ros" type="stageros" name="simulator" args="$(find TurboUlla)/stage/frobitpro_robolab_rsd.world"/>
		<remap from="/fmKnowledge/pose" to="/odom" />
		<!-- remap from="/base_pose_ground_truth" to="/fmInformation/gpgga_tranmerc" / -->
        </group>
</launch>
